### PR TITLE
Samples

### PR DESCRIPTION
**Pull request by TrevorFSmith**
This PR introduces a couple of samples:
- `single-page-app` shows how to add the babel plugin to an app created by `create-react-app`
- `styled-components` shows how to add the babel plugin to a very simple React app that uses styled-components

### CODE REVIEW

**Comment by TrevorFSmith** [comment]
@JoshMiers-FS 

The more I look at the [styled-component babel code](https://github.com/davesnx/babel-plugin-transform-react-qa-classes/blob/master/src/styled-components.js) from the original repo the more it looks to me like it's not doing anything we need so people could combine them with no problems. Maybe I'm missing something, though?

Here's the input from [this sample]():

	const Title = styled.h1`
	  font-family: sans-serif;
	  font-size: 1.5em;
	`;
	const Anchor = styled.a`
	  color: purple;
	`;
	class App extends Component {
	  render() {
	    return (
	      <div className="foo">
	        <Title>
	          A title
	          <br />
	          <Anchor href="#foo">An anchor</Anchor>
	        </Title>
	      </div>
	    );
	  }
	}

And this is the render:

	<div class="foo" data-element="div" data-component="App" data-source-file="App.js">
		<h1 data-element="Title" class="App__Title-sc-17 kmanzR">
			A title
			<br data-element="br">
			<a href="#foo" data-element="Anchor" class="App__Anchor-sc-17 dFfJuD">
				An anchor
			</a>
		</h1>
	</div>

It seems like the render has all of the useful information (for our customers) without the styled-components-specific code in our plugin.

What do you think?

**Comment by JoshMiers-FS** [comment]
I like that this is a use-case for `data-element="Anchor"`.

I agree based on the code you pasted. Are those the only types of styled-components declarations available?

Maybe the old styled-components code would have marked it as what we call `data-component`?

**Comment by JoshMiers-FS** [comment]
I am wondering if `data-component` would be more correct or if we should use `data-component` for custom _components_ only and utilize `data-element` for "swaps" like styled components is doing (and also for what react native does under the covers).

**Comment by TrevorFSmith** [comment]
@JoshMiers-FS

One can also create styled components like so:
    const Para = styled.p.attrs(() => ({ 'data-test': 'blatz' }))`
      border: solid 1px purple;
    `
    *snip*
    <Para>Test</Para>

The render is then like this:

    <p data-element="Para" data-test="blatz" class="App__Para-sc-17 kCYhrS">Test</p>

**Comment by TrevorFSmith** [comment]
> I am wondering if `data-component` would be more correct or if we should use `data-component` for custom _components_ only and utilize `data-element` for "swaps" like styled components is doing (and also for what react native does under the covers).

I kind of like having data-element for style-components since in use they feel more like elements than React.Components even though I know under the covers they're components.

**Comment by JoshMiers-FS** [comment]
> I kind of like having data-element for style-components since in use they feel more like elements than React.Components even though I know under the covers they're components.

I agree.

**Comment by patrick-fs** [comment]
For FullStory users who are tasked with finding components in search, how would they know to search for a `data-component` vs. a `data-element`?

**Comment by JoshMiers-FS** [comment]
> For FullStory users who are tasked with finding things in search, how would they know to search for a `data-component` vs. a `data-element`?

Right now it would be `data-component` only for `React.Components`. Anything else would come through as `data-element`.

**Comment by patrick-fs** [comment]
> > For FullStory users who are tasked with finding things in search, how would they know to search for a `data-component` vs. a `data-element`?
> 
> Right now it would be `data-component` only for `React.Components`. Anything else would come through as `data-element`.

I think we'd need to consider whether our users can make that distinction. If they can't then they'd need to be able to write queries along the lines of "clicked [data-component='Bananas'] or [data-element='Bananas']"

Having said all of that, @TrevorFSmith in a prior PR, I think all React components were decorated with a `data-component` attribute: https://github.com/fullstorydev/fullstory-babel-plugin-annotate-react/pull/2#issuecomment-632132873

Is this still the case (or maybe I misunderstood)?

**Comment by chiplay** [comment]
As long as the attributes are in the DOM and we have them indexed, it shouldn't matter too much if it's `data-component` or `data-element` from a search perspective, since autocomplete will match on the **values** of the attribute.  e.g. searching for `Anchor` will return `data-element="Anchor"`, etc.  Though I do see your point about them getting conflated if the same components shows up in two different attributes.

It's important to get this into some real world examples and stress test these ideas with customers so we can validate the approach and adjust if necessary.

**Comment by TrevorFSmith** [comment]
@patrick-fs wrote:

> @TrevorFSmith in a prior PR, you verified that nested React components were decorated with a `data-component` attribute: [#2 (comment)](https://github.com/fullstorydev/fullstory-babel-plugin-annotate-react/pull/2#issuecomment-632132873). Is this still the case?

Yes, the patterns for the usual React components decoration hasn't changed. They're always tagged with `data-component`. The only new information is that DOM elements created via styled-components are tagged with `data-element`. Combined with the just merged PR that eliminates the useless `data-element` attributes for the usual DOM elements, I think this makes `data-element` pretty useful. 

> Your example further up the thread tripped me up because a nested component was tagged with a `data-element` attribute alone. In this case I can see an issue where users won't know what to search for: `data-component` vs. `data-element`.

The logic is that if you use the React patterns for Components they'll be tagged with `data-component` and if you use libs like styled-components to create DOM then they'll be tagged with the variable name in `data-element`. The worst case is that the user inspects the element and chooses the right attribute when writing a selector.

File: README.md [hunk 1]
```diff
@@ -23,10 +23,10 @@ Example output:
     class HelloComponent extends Component {
       render() {
         return React.createElement("div", {
-          "data-element": "div",
-          "data-component": "HelloComponent"
+          "data-component": "HelloComponent",
+          "data-file-source": "hello-component.js"
         }, React.createElement("h1", {
-          "data-element": "h1"
+          null
```
Comment (patrick-fs) [inline README.md:29]: @TrevorFSmith I think it would be helpful to have an additional example of the rendered HTML as well.

**Comment by patrick-fs** [comment]
> The logic is that if you use the React patterns for Components they'll be tagged with `data-component` and if you use libs like styled-components to create DOM then they'll be tagged with the variable name in `data-element`. The worst case is that the user inspects the element and chooses the right attribute when writing a selector.

@TrevorFSmith Would it be a good idea to include this behavior in the README? Styled components still look like React components when they are composed together (can I coin the term "Duck Component"?) and describing this in the Readme could help build a clear mental model for our users ("A note about styled components"). [Your description](https://github.com/fullstorydev/fullstory-babel-plugin-annotate-react/pull/3#issuecomment-634799496) as well as the [example above](https://github.com/fullstorydev/fullstory-babel-plugin-annotate-react/pull/3#issuecomment-634135460) are perfect.

**Comment by TrevorFSmith** [comment]
@patrick-fs Ok, I've addressed your comments regarding the documentation so it's ready for review.

File: samples/styled-components/README.md [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+A browser window should open and you can compare the source components in `fullstory-babel-plugin-annotate-react/samples/styled-components/src/components/App.js` to the rendered DOM. Look for the `data-component`, `data-element`, and `data-source-file` attributes added by this plugin.
+
+## A note about attribute names
+
+If you use the usual React code patterns for Components then they'll be tagged with `data-component` and if you use libs like `styled-components` to create DOM then they'll be tagged with the variable name in `data-element`.
+
+For example, here are two styled components and a normal React.Component:
+
+	const Title = styled.h1`
```
Comment (patrick-fs) [inline samples/styled-components/README.md:21]: nit: please use spaces instead of tabs, 2 spaces per indentation

**Pull request merged**